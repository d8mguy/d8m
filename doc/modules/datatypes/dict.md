## Dict

Module dict provides dictionaries, aka associative arrays, as an abstract type. It also defines an STMap to go's map type. Dict uses extensional equality (written as ==) to look up keys. You can define a version of dict that uses === in a few lines of code if you need that.

    [dict(KT,VT): x0]
    [dict(KT,VT): kz, vz]

where `x0:list(pairT)`, `kz:list(KT)`, `vz:list(VT)`.

    val dict = \typegen(KT, VT::entity) {
        val pairT = tuple(key: KT, value: VT)
        extend tuple(...private...) where {
            method count = \() -> integer {...}
            method rvalindex = \(k:KT) -> nilPossible(VT) {...}
            method lvalindex = \mod(k:KT, v:VT) -> nothing {...}
            method remove = \mod(k:KT) -> nothing {...}
            method keys = \() -> list(KT) {...}
            method values = \() -> list(VT) {...}
        }
    }
