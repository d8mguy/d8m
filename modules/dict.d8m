// Module dict provides dictionaries, aka associative arrays, as an abstract type. It also defines an STMap to go's map type.
// Dict uses extensional equality (written as ==) to look up keys. You can define a version of dict that uses === in a few
// lines of code if you need that.

export dict

val dict = \typegen(KT, VT::entity) {
	val pairT = tuple(key: KT, value: VT)
	extend tuple(private x: list(pairT)) where {
		assert x.count == x.{key}.uniq.count
		assert enumerable(x)
		method lithook = \mod(x0: list(pairT)) { x = x0 }
		method lithook = \mod(kz: list(KT), vz: list(VT)) { x = (0...kz.count).{ [pairT: kz[this], vz[this]] } }
		method count = \() { x.count }
		method rvalindex = \(k: KT) { val tmp = x[key == k => index]; tmp == nil ? nil : x[tmp].value }
		method lvalindex = \mod(k: KT, v: VT) -> nothing {
			var tmp = x[key == k => index]
			if(tmp == nil) x.pushb([pairT: k, v]) else x[tmp].value = v
		}
		method remove = \mod(k: KT) {
			val inx = x[key == k => index]
			unless(inx == nil) x.removeIndex(inx)
		}
		method keys = \() { x.{key} }
		method values = \() { x.{value} }
	}
}
given(T1, T2::entity, aT1: T1, alpr: list(tuple(key: T1, value: T2))) {
	STMap {
		source = dict(T1, T2)
		target = Gomap(T1, T2)
		direct lvalindex, rvalindex, count
		self.remove(aT1) => self.delete(aT1)
		self.lithook(alpr) => self.lithook(alpr)
		// TBD: keys, values
	}
}
