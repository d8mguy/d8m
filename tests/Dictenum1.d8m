// dictenum1
import "dict" melted
val pair = tuple(a, b: integer)
val try = \mod(pr: pair) {
    var d1 = [dict(integer, pair):]
    d1[1] = pr
    d1[4] = copy(pr)
    pr.a = 12
    println(d1.keys)
    println(d1.values)
}
try([pair: 33, -5])
