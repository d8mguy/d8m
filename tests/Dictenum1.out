[1, 4]
[[12; -5], [33; -5]]
